src/main/java/mate/academy/springboot/datajpa/controller/ProductController.java

```diff
@@ -0,0 +1,59 @@
+package mate.academy.springboot.datajpa.controller;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import mate.academy.springboot.datajpa.dto.ProductRequestDto;
+import mate.academy.springboot.datajpa.dto.ProductResponseDto;
+import mate.academy.springboot.datajpa.dto.mapper.DtoMapper;
+import mate.academy.springboot.datajpa.model.Product;
+import mate.academy.springboot.datajpa.service.ProductService;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/products")
+public class ProductController {
+    private final ProductService productService;
+    private final DtoMapper<ProductRequestDto, ProductResponseDto, Product> productMapper;
+
+    @PostMapping
+    public ProductResponseDto add(@RequestBody ProductRequestDto productRequestDto) {
+        Product product = productService.add(productMapper.mapToModel(productRequestDto));
+        return productMapper.mapToDto(product);
+    }
+
+    @GetMapping("/{id}")
+    public ProductResponseDto get(@PathVariable Long id) {
+        return productMapper.mapToDto(productService.getById(id));
+    }
+
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id) {
+        productService.delete(id);
+    }
+
+    @PutMapping("/{id}")
+    public void update(@PathVariable Long id,
+                       @RequestBody ProductRequestDto productRequestDto) {
+        Product product = productMapper.mapToModel(productRequestDto);
+        product.setId(id);
+        productService.update(product);
+    }
+
+    @GetMapping("/products")
```

Comment:
your URL looks like `products/products/` - what is that?? Is it a user-friendly endpoint?

Reply:
ok
amended